File: api/_lib/editorial-mix-pool-schema.ts
/**
 * Editorial Mix Pool artifact contract. (Phase 3D-3C.1)
 *
 * A SEPARATE artifact from the raw Mix Pool. The raw pool (schema version 1) carries only
 * what the selector needs; this one adds the story-level editorial fields the iOS client
 * needs, so a selected five can become a `SignalsFeed` with no request-time generation.
 *
 * STORY-LEVEL vs EDITION-LEVEL — the central separation.
 *   Story-level fields belong to a candidate and are stable wherever it appears:
 *     headline, summary, keyTakeaways, whyItMatters, source, originalURL, readTime,
 *     imageURL, placeTime, audioURL, category.
 *   Edition-level fields depend on WHICH five were chosen and in what order:
 *     number, importance, lead, date, focus, version.
 * Storing an edition-level field on a reusable candidate would be a lie — the same story
 * is `number: 1` in one mix and `number: 4` in another. The schema therefore REJECTS
 * `number`, `importance` and `lead` inside a stored candidate, and `editorial-mix-feed.ts`
 * assigns them after selection.
 *
 * WHY `focus` IS CONSTANT. `pipeline/build.py` sets `FOCUS = "MIXED"` and `VERSION = 1` as
 * module constants and writes them verbatim; neither is derived from the selected stories
 * and neither needs a model call. A Custom Mix edition can use the same values.
 *
 * WHY LISTEN IS NOT A BLOCKER. In the live feed `audioURL` is `""`, and `listen` and
 * `localized` are optional in `SignalsFeed.swift`. A personalized edition decodes without
 * any audio artifact. Personalized Listen is a separate concern, deliberately out of scope.
 *
 * TWO IDENTITIES, DELIBERATELY.
 *   `selectorPoolIdentity`  hashes ONLY the extracted selector candidates, reusing the raw
 *                           Mix Pool identity function — so it must reproduce the original
 *                           raw pool's `poolIdentity` exactly. This proves enrichment did
 *                           not disturb selection inputs.
 *   `editorialPoolIdentity` hashes the whole enriched candidate set. Editorial drift moves
 *                           this one and leaves the selector identity untouched, which is
 *                           what makes the two failure modes distinguishable in diagnostics.
 *
 * All raw-candidate validation and numeric rules are DELEGATED to `mix-pool-schema.ts`.
 * There is no second copy of the numeric contract.
 *
 * Pure: no filesystem, network, environment, clock or logging. Not imported by any route,
 * the orchestrator or the runtime factory.
 */

import { createHash } from "node:crypto";

import {
  CANDIDATE_OPTIONAL,
  CANDIDATE_REQUIRED,
  canonicalMixPoolBytes,
  mixPoolIdentity,
  validateMixPoolArtifact,
  type JsonValue,
} from "./mix-pool-schema.js";

export const EDITORIAL_ARTIFACT_TYPE = "editorial-mix-pool";
export const EDITORIAL_SCHEMA_VERSION = 1;
export const EDITORIAL_SELECTOR_VERSION = 1;
export const EDITORIAL_VERSION = 1;
export const EDITORIAL_GENERATOR_VERSION = 1;

/** Future pipeline targets, recorded here so the contract and the plan cannot drift. */
export const TARGET_POOL_SIZE = 20;
export const MINIMUM_PUBLISHABLE_POOL_SIZE = 15;

export const EDITORIAL_TOP_KEYS = [
  "artifactType",
  "schemaVersion",
  "selectorVersion",
  "editorialVersion",
  "date",
  "generatedAt",
  "candidateCount",
  "selectorPoolIdentity",
  "editorialPoolIdentity",
  "candidates",
  "provenance",
] as const;

/** Story-level fields. Every one is required; none is edition-dependent. */
export const EDITORIAL_REQUIRED = [
  "category",
  "source",
  "headline",
  "summary",
  "keyTakeaways",
  "whyItMatters",
  "originalURL",
  "readTime",
  "imageURL",
  "placeTime",
  "audioURL",
] as const;

/** Never storable on a candidate: these depend on the chosen five. */
export const EDITION_LEVEL_FIELDS = ["number", "importance", "lead", "focus", "version"] as const;

/** Never storable anywhere: provider internals, prompts, bodies, local paths. */
export const EDITORIAL_FORBIDDEN_KEYS = [
  "prompt",
  "prompts",
  "rawbody",
  "raw_body",
  "articlebody",
  "article_body",
  "fulltext",
  "full_text",
  "sourcetext",
  "source_text",
  "providerresponse",
  "provider_response",
  "apikey",
  "api_key",
] as const;

const READ_TIME_MIN = 1;
const READ_TIME_MAX = 60;
const MAX_TAKEAWAYS = 8;

export type EditorialStory = {
  category: string;
  source: string;
  headline: string;
  summary: string;
  keyTakeaways: string[];
  whyItMatters: string;
  originalURL: string;
  readTime: number;
  imageURL: string;
  placeTime: string;
  audioURL: string;
};

export type EnrichedCandidate = {
  selector: Record<string, JsonValue>;
  editorial: EditorialStory;
};

export type EditorialValidationResult = { valid: boolean; errors: string[] };

function isObject(value: unknown): value is Record<string, JsonValue> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function nonEmpty(value: unknown): boolean {
  return typeof value === "string" && value.trim().length > 0;
}

function isHttpsUrl(value: unknown): boolean {
  if (typeof value !== "string") return false;
  try {
    const parsed = new URL(value);
    return parsed.protocol === "https:" && Boolean(parsed.host);
  } catch {
    return false;
  }
}

/** Extract the raw selector candidates, in stored order. */
export function extractSelectorCandidates(artifact: unknown): JsonValue[] {
  if (!isObject(artifact) || !Array.isArray(artifact.candidates)) return [];
  return (artifact.candidates as JsonValue[]).map((c) =>
    isObject(c) ? ((c.selector ?? {}) as JsonValue) : ({} as JsonValue),
  );
}

/** Reuses the RAW pool identity function, so it must equal the original pool's value. */
export function selectorPoolIdentityOf(artifact: unknown): string {
  return mixPoolIdentity(extractSelectorCandidates(artifact));
}

/** Hash of the whole enriched candidate set, sorted by selector id for stability. */
export function editorialPoolIdentityOf(candidates: JsonValue[]): string {
  const ordered = [...candidates].sort((a, b) => {
    const idA = String((a as { selector?: { id?: unknown } })?.selector?.id ?? "");
    const idB = String((b as { selector?: { id?: unknown } })?.selector?.id ?? "");
    return idA < idB ? -1 : idA > idB ? 1 : 0;
  });
  return createHash("sha256").update(canonicalMixPoolBytes(ordered)).digest("hex");
}

function forbiddenKeyPaths(value: JsonValue, path = "$"): string[] {
  const found: string[] = [];
  if (Array.isArray(value)) {
    value.forEach((child, i) => found.push(...forbiddenKeyPaths(child, `${path}[${i}]`)));
  } else if (isObject(value)) {
    for (const [key, child] of Object.entries(value)) {
      const childPath = `${path}.${key}`;
      if ((EDITORIAL_FORBIDDEN_KEYS as readonly string[]).includes(key.toLowerCase())) {
        found.push(childPath);
      }
      found.push(...forbiddenKeyPaths(child, childPath));
    }
  } else if (typeof value === "string" && (value.startsWith("file://") || value.includes("/Users/"))) {
    found.push(path);
  }
  return found;
}

/** Validate one story-level block. Errors carry field paths only — never content. */
export function validateEditorialStory(story: unknown, prefix: string, errors: string[]): void {
  if (!isObject(story)) {
    errors.push(`${prefix} must be an object`);
    return;
  }
  const present = new Set(Object.keys(story));
  for (const key of [...EDITORIAL_REQUIRED].filter((k) => !present.has(k)).sort()) {
    errors.push(`${prefix} missing field: ${key}`);
  }
  for (const key of [...present].filter(
    (k) => !(EDITORIAL_REQUIRED as readonly string[]).includes(k),
  ).sort()) {
    errors.push(`${prefix} unsupported field: ${key}`);
  }

  for (const field of ["category", "source", "headline", "summary", "whyItMatters", "placeTime"] as const) {
    if (!nonEmpty(story[field])) errors.push(`${prefix}.${field} must be nonempty`);
  }

  const takeaways = story.keyTakeaways;
  if (!Array.isArray(takeaways) || takeaways.length === 0) {
    errors.push(`${prefix}.keyTakeaways must be a nonempty array`);
  } else {
    if (takeaways.length > MAX_TAKEAWAYS) {
      errors.push(`${prefix}.keyTakeaways exceeds ${MAX_TAKEAWAYS} entries`);
    }
    takeaways.forEach((item, index) => {
      if (!nonEmpty(item)) errors.push(`${prefix}.keyTakeaways[${index}] must be nonempty`);
    });
  }

  // `build.py` requires a real https article URL; mirror that exactly.
  if (!isHttpsUrl(story.originalURL)) {
    errors.push(`${prefix}.originalURL must be an https article URL`);
  }
  if (!isHttpsUrl(story.imageURL)) {
    errors.push(`${prefix}.imageURL must be an https URL`);
  }

  const readTime = story.readTime;
  if (typeof readTime === "boolean" || !Number.isSafeInteger(readTime)) {
    errors.push(`${prefix}.readTime must be an integer`);
  } else if ((readTime as number) < READ_TIME_MIN || (readTime as number) > READ_TIME_MAX) {
    errors.push(`${prefix}.readTime must be within [${READ_TIME_MIN}, ${READ_TIME_MAX}]`);
  }

  // v1 compatibility: the live feed carries "" here.
  if (typeof story.audioURL !== "string") errors.push(`${prefix}.audioURL must be a string`);
}

/** Validate the whole Editorial Mix Pool artifact. */
export function validateEditorialMixPool(artifact: unknown): EditorialValidationResult {
  const errors: string[] = [];
  if (!isObject(artifact)) return { valid: false, errors: ["artifact must be an object"] };

  const keys = new Set(Object.keys(artifact));
  for (const key of [...EDITORIAL_TOP_KEYS].filter((k) => !keys.has(k)).sort()) {
    errors.push(`missing top-level field: ${key}`);
  }
  for (const key of [...keys].filter(
    (k) => !(EDITORIAL_TOP_KEYS as readonly string[]).includes(k),
  ).sort()) {
    errors.push(`unsupported top-level field: ${key}`);
  }

  if (artifact.artifactType !== EDITORIAL_ARTIFACT_TYPE) errors.push("unsupported artifactType");
  if (artifact.schemaVersion !== EDITORIAL_SCHEMA_VERSION) errors.push("unsupported schemaVersion");
  if (artifact.selectorVersion !== EDITORIAL_SELECTOR_VERSION) {
    errors.push("unsupported selectorVersion");
  }
  if (artifact.editorialVersion !== EDITORIAL_VERSION) errors.push("unsupported editorialVersion");

  if (typeof artifact.date !== "string" || !/^\d{4}-\d{2}-\d{2}$/.test(artifact.date)) {
    errors.push("date must be ISO YYYY-MM-DD");
  }
  if (!nonEmpty(artifact.generatedAt) || !Number.isFinite(Date.parse(String(artifact.generatedAt)))) {
    errors.push("generatedAt must be an ISO datetime");
  }

  const candidates = Array.isArray(artifact.candidates) ? (artifact.candidates as JsonValue[]) : [];
  if (!Array.isArray(artifact.candidates)) errors.push("candidates must be an array");
  if (artifact.candidateCount !== candidates.length) {
    errors.push("candidateCount does not match candidates");
  }

  const ids = new Set<string>();
  const urls = new Set<string>();
  candidates.forEach((raw, index) => {
    const prefix = `candidates[${index}]`;
    if (!isObject(raw)) {
      errors.push(`${prefix} must be an object`);
      return;
    }
    const shape = Object.keys(raw).sort().join(",");
    if (shape !== "editorial,selector") {
      errors.push(`${prefix} must contain exactly selector and editorial`);
    }

    const selector = raw.selector;
    if (!isObject(selector)) {
      errors.push(`${prefix}.selector must be an object`);
    } else {
      // Edition-level fields must never be stored on a reusable candidate.
      for (const field of EDITION_LEVEL_FIELDS) {
        if (field in selector) errors.push(`${prefix}.selector must not carry ${field}`);
      }
      const id = selector.id;
      if (!nonEmpty(id)) errors.push(`${prefix}.selector.id must be nonempty`);
      else if (ids.has(id as string)) errors.push(`duplicate candidate id: ${id}`);
      else ids.add(id as string);

      const url = selector.url;
      if (typeof url === "string" && url.length > 0) {
        if (urls.has(url)) errors.push(`duplicate candidate url`);
        else urls.add(url);
      }
    }

    const editorial = raw.editorial;
    if (!isObject(editorial)) {
      errors.push(`${prefix}.editorial must be an object`);
    } else {
      for (const field of EDITION_LEVEL_FIELDS) {
        if (field in editorial) errors.push(`${prefix}.editorial must not carry ${field}`);
      }
      validateEditorialStory(editorial, `${prefix}.editorial`, errors);

      // The editorial block must not contradict the selector's taxonomy.
      if (isObject(selector) && editorial.category !== selector.category) {
        errors.push(`${prefix}.editorial.category does not match the selector category`);
      }
      if (isObject(selector) && nonEmpty(selector.source) && editorial.source !== selector.source) {
        errors.push(`${prefix}.editorial.source does not match the selector source`);
      }
    }
  });

  // Delegate ALL raw candidate rules — numeric contract, taxonomy, evidence — to the raw
  // schema by validating a synthetic raw artifact built from the extracted selectors.
  const selectors = extractSelectorCandidates(artifact);
  if (selectors.length > 0) {
    const rawProbe = {
      schemaVersion: 1,
      selectorVersion: 1,
      date: artifact.date,
      generatedAt: artifact.generatedAt,
      poolIdentity: mixPoolIdentity(selectors),
      candidateCount: selectors.length,
      candidates: selectors,
      validation: { valid: true, errors: [], warnings: [] },
      provenance: {
        source: "editorial-mix-pool",
        inputIdentity: "0".repeat(64),
        generatorVersion: 1,
      },
    } as unknown as JsonValue;
    for (const error of validateMixPoolArtifact(rawProbe).errors) {
      // Skip envelope-only complaints that belong to the raw artifact, not to a candidate.
      if (error.startsWith("candidates[") || error.startsWith("duplicate ")) {
        errors.push(`selector: ${error}`);
      }
    }
  }

  if (artifact.selectorPoolIdentity !== mixPoolIdentity(selectors)) {
    errors.push("selectorPoolIdentity does not match the extracted selector candidates");
  }
  if (artifact.editorialPoolIdentity !== editorialPoolIdentityOf(candidates)) {
    errors.push("editorialPoolIdentity does not match the enriched candidates");
  }

  const provenance = artifact.provenance;
  if (!isObject(provenance)) {
    errors.push("provenance must be an object");
  } else {
    if (!nonEmpty(provenance.source)) errors.push("provenance.source must be nonempty");
    if (provenance.generatorVersion !== EDITORIAL_GENERATOR_VERSION) {
      errors.push("unsupported provenance.generatorVersion");
    }
  }

  for (const path of forbiddenKeyPaths(artifact as JsonValue)) {
    errors.push(`forbidden field or local path: ${path}`);
  }

  return { valid: errors.length === 0, errors };
}

/** Canonical bytes for the enriched artifact, using the proven raw-pool rules. */
export function canonicalEditorialBytes(artifact: JsonValue): Buffer {
  return canonicalMixPoolBytes(artifact);
}

export { CANDIDATE_REQUIRED, CANDIDATE_OPTIONAL };

File: pipeline/editorial_mix_pool_schema.py
#!/usr/bin/env python3
"""Editorial Mix Pool artifact contract. (Phase 3D-3C.1)

A SEPARATE artifact from the raw Mix Pool. Raw schema version 1 stays untouched; this adds
the story-level editorial fields the iOS client needs, so a selected five can become a
`SignalsFeed` with no request-time generation.

STORY-LEVEL vs EDITION-LEVEL. Story-level fields belong to a candidate and are stable
wherever it appears (headline, summary, keyTakeaways, whyItMatters, source, originalURL,
readTime, imageURL, placeTime, audioURL, category). Edition-level fields depend on WHICH
five were chosen and in what order (number, importance, lead, date, focus, version) and are
therefore REJECTED inside a stored candidate — the same story is number 1 in one mix and
number 4 in another.

`focus` is constant: `build.py` sets `FOCUS = "MIXED"` and `VERSION = 1` as module
constants, neither derived from the selection. Listen is not a decoding blocker: `audioURL`
is `""` in the live feed and `listen`/`localized` are optional in Swift.

TWO IDENTITIES. `selectorPoolIdentity` hashes only the extracted selector candidates,
reusing `mix_pool_schema.pool_identity`, so it must reproduce the original raw pool value —
proof that enrichment did not disturb selection inputs. `editorialPoolIdentity` hashes the
whole enriched set, so editorial drift is distinguishable from selector drift.

All raw candidate rules — the numeric contract, taxonomy, evidence — are DELEGATED to
`mix_pool_schema`. There is no second copy.

This module defines the contract only. It does not fetch articles, run the writer, assign
images or publish anything; that is Phase 3D-3C.2. Future targets: pool size 20, minimum
publishable count 15, images assigned offline with no duplicate inside the pool and the
existing 90-edition cooldown retained.
"""

from __future__ import annotations

import hashlib
from typing import Any
from urllib.parse import urlparse

try:
    from .mix_pool_schema import canonical_bytes, pool_identity, validate_artifact
except ImportError:  # direct script/module use, matching the rest of the pipeline
    from mix_pool_schema import (  # type: ignore[no-redef]
        canonical_bytes,
        pool_identity,
        validate_artifact,
    )

ARTIFACT_TYPE = "editorial-mix-pool"
SCHEMA_VERSION = 1
SELECTOR_VERSION = 1
EDITORIAL_VERSION = 1
GENERATOR_VERSION = 1

#: Future pipeline targets, recorded beside the contract so the two cannot drift.
TARGET_POOL_SIZE = 20
MINIMUM_PUBLISHABLE_POOL_SIZE = 15

TOP_KEYS = {
    "artifactType", "schemaVersion", "selectorVersion", "editorialVersion",
    "date", "generatedAt", "candidateCount", "selectorPoolIdentity",
    "editorialPoolIdentity", "candidates", "provenance",
}
EDITORIAL_REQUIRED = {
    "category", "source", "headline", "summary", "keyTakeaways", "whyItMatters",
    "originalURL", "readTime", "imageURL", "placeTime", "audioURL",
}
EDITION_LEVEL_FIELDS = {"number", "importance", "lead", "focus", "version"}
FORBIDDEN_KEYS = {
    "prompt", "prompts", "rawbody", "raw_body", "articlebody", "article_body",
    "fulltext", "full_text", "sourcetext", "source_text", "providerresponse",
    "provider_response", "apikey", "api_key",
}

READ_TIME_MIN = 1
READ_TIME_MAX = 60
MAX_TAKEAWAYS = 8


def _nonempty(value: Any) -> bool:
    return isinstance(value, str) and bool(value.strip())


def _https_url(value: Any) -> bool:
    if not isinstance(value, str):
        return False
    parsed = urlparse(value)
    return parsed.scheme == "https" and bool(parsed.netloc)


def extract_selector_candidates(artifact: Any) -> list[dict[str, Any]]:
    """The raw selector candidates, in stored order."""
    if not isinstance(artifact, dict) or not isinstance(artifact.get("candidates"), list):
        return []
    return [
        (row.get("selector") or {}) if isinstance(row, dict) else {}
        for row in artifact["candidates"]
    ]


def selector_pool_identity(artifact: Any) -> str:
    """Reuses the RAW identity function, so it equals the original pool's `poolIdentity`."""
    return pool_identity(extract_selector_candidates(artifact))


def editorial_pool_identity(candidates: list[dict[str, Any]]) -> str:
    """Hash of the whole enriched set, ordered by selector id for stability."""
    ordered = sorted(
        candidates,
        key=lambda row: str((row.get("selector") or {}).get("id", ""))
        if isinstance(row, dict) else "",
    )
    return hashlib.sha256(canonical_bytes(ordered)).hexdigest()


def _forbidden_paths(value: Any, path: str = "$") -> list[str]:
    found: list[str] = []
    if isinstance(value, dict):
        for key, child in value.items():
            child_path = f"{path}.{key}"
            if str(key).lower() in FORBIDDEN_KEYS:
                found.append(child_path)
            found.extend(_forbidden_paths(child, child_path))
    elif isinstance(value, list):
        for index, child in enumerate(value):
            found.extend(_forbidden_paths(child, f"{path}[{index}]"))
    elif isinstance(value, str) and (value.startswith("file://") or "/Users/" in value):
        found.append(path)
    return found


def validate_editorial_story(story: Any, prefix: str, errors: list[str]) -> None:
    """Validate one story-level block. Errors carry field paths only — never content."""
    if not isinstance(story, dict):
        errors.append(f"{prefix} must be an object")
        return

    for key in sorted(EDITORIAL_REQUIRED - set(story)):
        errors.append(f"{prefix} missing field: {key}")
    for key in sorted(set(story) - EDITORIAL_REQUIRED):
        errors.append(f"{prefix} unsupported field: {key}")

    for field in ("category", "source", "headline", "summary", "whyItMatters", "placeTime"):
        if not _nonempty(story.get(field)):
            errors.append(f"{prefix}.{field} must be nonempty")

    takeaways = story.get("keyTakeaways")
    if not isinstance(takeaways, list) or not takeaways:
        errors.append(f"{prefix}.keyTakeaways must be a nonempty array")
    else:
        if len(takeaways) > MAX_TAKEAWAYS:
            errors.append(f"{prefix}.keyTakeaways exceeds {MAX_TAKEAWAYS} entries")
        for index, item in enumerate(takeaways):
            if not _nonempty(item):
                errors.append(f"{prefix}.keyTakeaways[{index}] must be nonempty")

    if not _https_url(story.get("originalURL")):
        errors.append(f"{prefix}.originalURL must be an https article URL")
    if not _https_url(story.get("imageURL")):
        errors.append(f"{prefix}.imageURL must be an https URL")

    read_time = story.get("readTime")
    if isinstance(read_time, bool) or not isinstance(read_time, int):
        errors.append(f"{prefix}.readTime must be an integer")
    elif not (READ_TIME_MIN <= read_time <= READ_TIME_MAX):
        errors.append(f"{prefix}.readTime must be within [{READ_TIME_MIN}, {READ_TIME_MAX}]")

    if not isinstance(story.get("audioURL"), str):
        errors.append(f"{prefix}.audioURL must be a string")


def validate_editorial_mix_pool(artifact: Any) -> dict[str, Any]:
    """Validate the whole Editorial Mix Pool artifact."""
    errors: list[str] = []
    if not isinstance(artifact, dict):
        return {"valid": False, "errors": ["artifact must be an object"]}

    for key in sorted(TOP_KEYS - set(artifact)):
        errors.append(f"missing top-level field: {key}")
    for key in sorted(set(artifact) - TOP_KEYS):
        errors.append(f"unsupported top-level field: {key}")

    if artifact.get("artifactType") != ARTIFACT_TYPE:
        errors.append("unsupported artifactType")
    if artifact.get("schemaVersion") != SCHEMA_VERSION:
        errors.append("unsupported schemaVersion")
    if artifact.get("selectorVersion") != SELECTOR_VERSION:
        errors.append("unsupported selectorVersion")
    if artifact.get("editorialVersion") != EDITORIAL_VERSION:
        errors.append("unsupported editorialVersion")

    date = artifact.get("date")
    if not isinstance(date, str) or len(date) != 10 or date[4] != "-" or date[7] != "-":
        errors.append("date must be ISO YYYY-MM-DD")
    if not _nonempty(artifact.get("generatedAt")):
        errors.append("generatedAt must be an ISO datetime")

    candidates = artifact.get("candidates")
    if not isinstance(candidates, list):
        errors.append("candidates must be an array")
        candidates = []
    if artifact.get("candidateCount") != len(candidates):
        errors.append("candidateCount does not match candidates")

    ids: set[str] = set()
    urls: set[str] = set()
    for index, row in enumerate(candidates):
        prefix = f"candidates[{index}]"
        if not isinstance(row, dict):
            errors.append(f"{prefix} must be an object")
            continue
        if set(row) != {"selector", "editorial"}:
            errors.append(f"{prefix} must contain exactly selector and editorial")

        selector = row.get("selector")
        if not isinstance(selector, dict):
            errors.append(f"{prefix}.selector must be an object")
            selector = {}
        else:
            for field in sorted(EDITION_LEVEL_FIELDS & set(selector)):
                errors.append(f"{prefix}.selector must not carry {field}")
            identifier = selector.get("id")
            if not _nonempty(identifier):
                errors.append(f"{prefix}.selector.id must be nonempty")
            elif identifier in ids:
                errors.append(f"duplicate candidate id: {identifier}")
            else:
                ids.add(identifier)
            url = selector.get("url")
            if _nonempty(url):
                if url in urls:
                    errors.append("duplicate candidate url")
                else:
                    urls.add(url)

        editorial = row.get("editorial")
        if not isinstance(editorial, dict):
            errors.append(f"{prefix}.editorial must be an object")
        else:
            for field in sorted(EDITION_LEVEL_FIELDS & set(editorial)):
                errors.append(f"{prefix}.editorial must not carry {field}")
            validate_editorial_story(editorial, f"{prefix}.editorial", errors)
            if editorial.get("category") != selector.get("category"):
                errors.append(f"{prefix}.editorial.category does not match the selector category")
            if _nonempty(selector.get("source")) and editorial.get("source") != selector.get("source"):
                errors.append(f"{prefix}.editorial.source does not match the selector source")

    # Delegate every raw candidate rule to the raw schema via a synthetic raw artifact.
    selectors = extract_selector_candidates(artifact)
    if selectors:
        probe = {
            "schemaVersion": 1,
            "selectorVersion": 1,
            "date": artifact.get("date"),
            "generatedAt": artifact.get("generatedAt"),
            "poolIdentity": pool_identity(selectors),
            "candidateCount": len(selectors),
            "candidates": selectors,
            "validation": {"valid": True, "errors": [], "warnings": []},
            "provenance": {
                "source": "editorial-mix-pool",
                "inputIdentity": "0" * 64,
                "generatorVersion": 1,
            },
        }
        for error in validate_artifact(probe)["errors"]:
            if error.startswith("candidates[") or error.startswith("duplicate "):
                errors.append(f"selector: {error}")

    if artifact.get("selectorPoolIdentity") != pool_identity(selectors):
        errors.append("selectorPoolIdentity does not match the extracted selector candidates")
    if artifact.get("editorialPoolIdentity") != editorial_pool_identity(candidates):
        errors.append("editorialPoolIdentity does not match the enriched candidates")

    provenance = artifact.get("provenance")
    if not isinstance(provenance, dict):
        errors.append("provenance must be an object")
    else:
        if not _nonempty(provenance.get("source")):
            errors.append("provenance.source must be nonempty")
        if provenance.get("generatorVersion") != GENERATOR_VERSION:
            errors.append("unsupported provenance.generatorVersion")

    for path in _forbidden_paths(artifact):
        errors.append(f"forbidden field or local path: {path}")

    return {"valid": not errors, "errors": errors}


def freeze_editorial_artifact(
    *,
    date: str,
    generated_at: str,
    candidates: list[dict[str, Any]],
    source: str,
) -> dict[str, Any]:
    """Assemble a canonical Editorial Mix Pool artifact from enriched candidates."""
    ordered = sorted(
        candidates,
        key=lambda row: str((row.get("selector") or {}).get("id", "")),
    )
    artifact = {
        "artifactType": ARTIFACT_TYPE,
        "schemaVersion": SCHEMA_VERSION,
        "selectorVersion": SELECTOR_VERSION,
        "editorialVersion": EDITORIAL_VERSION,
        "date": date,
        "generatedAt": generated_at,
        "candidateCount": len(ordered),
        "selectorPoolIdentity": pool_identity(
            [(row.get("selector") or {}) for row in ordered]
        ),
        "editorialPoolIdentity": editorial_pool_identity(ordered),
        "candidates": ordered,
        "provenance": {"source": source, "generatorVersion": GENERATOR_VERSION},
    }
    return artifact
